### DISCUSSION TITLE
g_instanceDic 单例对象池存在的意义

### DISCUSSION THREAD

**Discussion by PApatigers** [Tencent/MMKV#732] [General]
目前MMKV 底层始终维护g_instanceDic  单例对象池，缓存所有已经注册过的mmkv实例对象
但由于注册阶段为保证g_instanceDic  的线程安全，所以开辟了g_instanceLock 单例锁，直到实例初始化完成才会释放，
这里会由于某个业务而阻塞其他所有业务的注册，且随着该对象的数据膨胀，这种阻塞会更加频繁。
想问下这里的单例对象池存在的意义，对应文件fd 应该交由上层业务主动管理open , close

**Comment by lingol** [comment]
It's actually more about the tradition of a KV storage engine. 
KV storage engines generally prefer a use-on-the-fly style. Users just get a KV store by ID, and do get/set operations on it, leaving lifetime management to the engine itself.

**Reply by PApatigers** [reply]
所有这种阻塞瓶颈是存在的，而且由于是持久化，所有会随着迭代时间的数据膨胀而增大概率，另一方面mmkv这里其实对于周期的管理也只是管理open，对于close还是需要业务主动管理

**Reply by lingol** [reply]
Actually, it's very rare for users to call `close()`. 
If you do come across the latency due to the lock of g_instanceDic, which I don't think it's common, you can always bypass it by caching the MMKV instance in your most frequency code.
